File: AlgoExpert/1. Easy/Java/8. Binary Search/Binary Search [O(log(n))].java
class Program {

	// O(log(n)) time | O(log(n)) space
	
	public static int binarySearch(int[] array, int target) {
		return binarySearch(array, target, 0, array.length - 1);
	}

	public static int binarySearch(int[] array, int target, int left, int right) {
		if (left > right) {
			return -1;
		}
		int middle = (left + right) / 2;
		int potentialMatch = array[middle];
		if (target == potentialMatch) {
			return middle;
		} else if (target < potentialMatch) {
			return binarySearch(array, target, left, middle - 1);
		} else {
		return binarySearch(array, target, middle + 1, right);
		}
	}
}

File: AlgoExpert/1. Easy/Python/8. Binary Search/Binary Search [O(log(n))].py
# O(log(n)) time | O(log(n)) space

def binarySearch(array, target):
	return binarySearchHelper(array, target, 0, len(array) - 1)

def binarySearchHelper(array, target, left, right):
	if left > right:
		return -1
	middle = (left + right) // 2
	potentialMatch = array[middle]
	if target == potentialMatch:
		return middle
	elif target < potentialMatch:
		return binarySearchHelper(array, target, left, middle - 1)
	else:
		return binarySearchHelper(array, target, middle + 1, right)